src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/ShortCircuitingRemoteServiceProvider.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.Razor.Remote;
+using Microsoft.CodeAnalysis.Remote.Razor;
+using Microsoft.ServiceHub.Framework;
+using Nerdbank.Streams;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.Test.Cohost;
+
+/// <summary>
+/// An implementation of IRemoteServiceProvider that doesn't actually do anything remote, but rather directly calls service methods
+/// </summary>
+internal class ShortCircuitingRemoteServiceProvider(ITestOutputHelper testOutputHelper) : IRemoteServiceProvider
+{
+    private static Dictionary<Type, IServiceHubServiceFactory> _factoryMap = BuildFactoryMap();
+
+    private readonly IServiceProvider _serviceProvider = new TestTraceSourceProvider(testOutputHelper);
+
+    private static Dictionary<Type, IServiceHubServiceFactory> BuildFactoryMap()
+    {
+        var result = new Dictionary<Type, IServiceHubServiceFactory>();
+
+        foreach (var type in typeof(RazorServiceFactoryBase<>).Assembly.GetTypes())
+        {
+            if (!type.IsAbstract &&
+                typeof(IServiceHubServiceFactory).IsAssignableFrom(type))
+            {
+                Debug.Assert(type.BaseType.GetGenericTypeDefinition() == typeof(RazorServiceFactoryBase<>));
```

Comment:
Since this is test infrastructure, consider just using an Xunit assert rather than a debug-only assert. Personally, I'd like to see this fail when running tests compiled as "release".